File: engine/controller.py
"""
Game Controller - orchestrates game phases and exposes APIs.

The controller manages the game lifecycle without any I/O operations.
It provides a clean interface for:
- Creating new games
- Processing turns
- Querying game state

All operations return new state snapshots and event logs.
"""

from __future__ import annotations

from engine.types import (
    PlayerId,
    LocationIndex,
    TurnNumber,
    InstanceId,
    Energy,
    Power,
    GamePhase,
    GameResult,
    STARTING_HAND_SIZE,
)
from engine.models import (
    GameState,
    PlayerState,
    CardInstance,
    CardDef,
    PlayerAction,
    create_initial_locations,
)
from engine.events import (
    GameEvent,
    GameStartedEvent,
    GameEndedEvent,
    TurnStartedEvent,
    TurnEndedEvent,
    EnergySetEvent,
    CardDrawnEvent,
)
from engine.rules import (
    resolve_actions,
    compute_winner,
    compute_location_powers,
    is_game_over,
)
from engine.cards import get_starter_deck_defs


class GameController:
    """
    Orchestrates game phases and provides APIs for game interaction.

    The controller is stateless - all state is passed in and returned.
    This makes it easy to:
    - Test with specific states
    - Implement undo/redo
    - Serialize game states
    - Replay games from event logs

    Usage:
        controller = GameController()
        state, events = controller.create_game()
        state, events = controller.start_turn(state)
        state, events = controller.resolve_turn(state, action_p0, action_p1)
    """

    def create_game(
        self,
        deck_defs_p0: tuple[CardDef, ...] | None = None,
        deck_defs_p1: tuple[CardDef, ...] | None = None,
    ) -> tuple[GameState, list[GameEvent]]:
        """
        Create a new game with initial state.

        Args:
            deck_defs_p0: Card definitions for player 0's deck (uses default if None)
            deck_defs_p1: Card definitions for player 1's deck (uses default if None)

        Returns:
            Tuple of (initial_state, events)
        """
        events: list[GameEvent] = []

        # Use default decks if not provided
        if deck_defs_p0 is None:
            deck_defs_p0 = get_starter_deck_defs()
        if deck_defs_p1 is None:
            deck_defs_p1 = get_starter_deck_defs()

        # Create card instances
        next_id = InstanceId(0)

        def create_deck(
            defs: tuple[CardDef, ...], player_id: PlayerId
        ) -> tuple[tuple[CardInstance, ...], InstanceId]:
            nonlocal next_id
            cards: list[CardInstance] = []
            for card_def in defs:
                card = CardInstance(
                    instance_id=next_id,
                    card_def=card_def,
                    owner=player_id,
                    permanent_power_modifier=Power(0),
                    ongoing_power_modifier=Power(0),
                    revealed=False,
                )
                cards.append(card)
                next_id = InstanceId(next_id + 1)
            return tuple(cards), next_id

        deck_p0, next_id = create_deck(deck_defs_p0, PlayerId(0))
        deck_p1, next_id = create_deck(deck_defs_p1, PlayerId(1))

        # Create player states with empty hands (will draw during turn start)
        player_0 = PlayerState(
            player_id=PlayerId(0),
            deck=deck_p0,
            hand=(),
            energy=Energy(0),
            max_energy=Energy(0),
        )
        player_1 = PlayerState(
            player_id=PlayerId(1),
            deck=deck_p1,
            hand=(),
            energy=Energy(0),
            max_energy=Energy(0),
        )

        # Create initial game state
        state = GameState(
            turn=TurnNumber(0),  # Will be set to 1 on first start_turn
            phase=GamePhase.TURN_START,
            players=(player_0, player_1),
            locations=create_initial_locations(),
            result=GameResult.IN_PROGRESS,
            next_instance_id=next_id,
        )

        events.append(GameStartedEvent())

        # Draw initial hands
        for player_id in (PlayerId(0), PlayerId(1)):
            for _ in range(STARTING_HAND_SIZE):
                player = state.get_player(player_id)
                new_player, drawn_card = player.draw_card()
                state = state.with_player(player_id, new_player)

                if drawn_card is not None:
                    events.append(
                        CardDrawnEvent(
                            player_id=player_id,
                            card_instance_id=drawn_card.instance_id,
                        )
                    )

        return state, events

    def start_turn(self, state: GameState) -> tuple[GameState, list[GameEvent]]:
        """
        Start a new turn.

        - Increments turn number
        - Clears turn-specific tracking (moved cards)
        - Sets energy to turn number
        - Draws a card for each player

        Args:
            state: Current game state

        Returns:
            Tuple of (new_state, events)
        """
        events: list[GameEvent] = []

        # Increment turn
        new_turn = TurnNumber(state.turn + 1)
        state = state.with_turn(new_turn)

        # Clear turn-specific tracking (e.g., cards moved this turn)
        state = state.clear_turn_tracking()

        events.append(TurnStartedEvent(turn=new_turn))

        # Set energy and draw for each player
        for player_id in (PlayerId(0), PlayerId(1)):
            player = state.get_player(player_id)

            # Set energy = turn number
            new_energy = Energy(new_turn)
            new_player = player.with_energy(new_energy, new_energy)
            state = state.with_player(player_id, new_player)

            events.append(
                EnergySetEvent(
                    player_id=player_id,
                    energy=new_energy,
                )
            )

            # Draw a card (if not first turn - first turn already drew 3)
            if new_turn > TurnNumber(1):
                player = state.get_player(player_id)
                new_player, drawn_card = player.draw_card()
                state = state.with_player(player_id, new_player)

                if drawn_card is not None:
                    events.append(
                        CardDrawnEvent(
                            player_id=player_id,
                            card_instance_id=drawn_card.instance_id,
                        )
                    )

        state = state.with_phase(GamePhase.PLANNING)

        return state, events

    def resolve_turn(
        self,
        state: GameState,
        action_p0: PlayerAction,
        action_p1: PlayerAction,
    ) -> tuple[GameState, list[GameEvent]]:
        """
        Resolve both players' actions and end the turn.

        Args:
            state: Current game state (should be in PLANNING phase)
            action_p0: Player 0's action
            action_p1: Player 1's action

        Returns:
            Tuple of (new_state, events)
        """
        events: list[GameEvent] = []

        # Set phase to resolution
        state = state.with_phase(GamePhase.RESOLUTION)

        # Resolve actions
        state, action_events = resolve_actions(state, action_p0, action_p1)
        events.extend(action_events)

        # End turn
        events.append(TurnEndedEvent(turn=state.turn))
        state = state.with_phase(GamePhase.TURN_END)

        # Check for game end
        if is_game_over(state):
            result, location_winners = compute_winner(state)
            location_powers = compute_location_powers(state)
            total_power = (
                Power(sum(p[0] for p in location_powers)),
                Power(sum(p[1] for p in location_powers)),
            )

            state = state.with_result(result)
            state = state.with_phase(GamePhase.GAME_OVER)

            events.append(
                GameEndedEvent(
                    result=result,
                    location_winners=location_winners,
                    location_powers=location_powers,
                    total_power=total_power,
                )
            )

        return state, events

    def step(
        self,
        state: GameState,
        action_p0: PlayerAction,
        action_p1: PlayerAction,
    ) -> tuple[GameState, list[GameEvent]]:
        """
        Execute a complete turn step: start turn (if needed), resolve actions.

        This is a convenience method that handles the full turn cycle.
        Useful for game loops that don't need fine-grained phase control.

        Args:
            state: Current game state
            action_p0: Player 0's action
            action_p1: Player 1's action

        Returns:
            Tuple of (new_state, all_events)
        """
        all_events: list[GameEvent] = []

        # Start turn if needed
        if state.phase == GamePhase.TURN_START or state.phase == GamePhase.TURN_END:
            state, turn_events = self.start_turn(state)
            all_events.extend(turn_events)

        # Resolve turn
        state, resolve_events = self.resolve_turn(state, action_p0, action_p1)
        all_events.extend(resolve_events)

        return state, all_events

    def get_legal_actions(
        self,
        state: GameState,
        player_id: PlayerId,
    ) -> list[PlayerAction]:
        """
        Get all legal actions for a player.

        Args:
            state: Current game state
            player_id: The player to get actions for

        Returns:
            List of legal PlayerAction objects
        """
        from engine.models import PassAction, PlayCardAction
        from engine.rules import validate_action

        actions: list[PlayerAction] = []

        # Pass is always legal
        actions.append(PassAction(player_id=player_id))

        # Try each card in hand at each location
        player = state.get_player(player_id)
        for card in player.hand:
            for loc_idx in (LocationIndex(0), LocationIndex(1), LocationIndex(2)):
                action = PlayCardAction(
                    player_id=player_id,
                    card_instance_id=card.instance_id,
                    location=loc_idx,
                )
                result = validate_action(state, action)
                if result.valid:
                    actions.append(action)

        return actions


# =============================================================================
# Utility Functions
# =============================================================================


def create_test_state(
    turn: int = 1,
    p0_energy: int = 1,
    p1_energy: int = 1,
    p0_hand_defs: tuple[CardDef, ...] = (),
    p1_hand_defs: tuple[CardDef, ...] = (),
) -> GameState:
    """
    Create a GameState for testing purposes.

    This is a convenience function for tests that need specific states.
    """
    next_id = InstanceId(0)

    def create_hand(
        defs: tuple[CardDef, ...], player_id: PlayerId
    ) -> tuple[tuple[CardInstance, ...], InstanceId]:
        nonlocal next_id
        cards: list[CardInstance] = []
        for card_def in defs:
            card = CardInstance(
                instance_id=next_id,
                card_def=card_def,
                owner=player_id,
                permanent_power_modifier=Power(0),
                ongoing_power_modifier=Power(0),
                revealed=False,
            )
            cards.append(card)
            next_id = InstanceId(next_id + 1)
        return tuple(cards), next_id

    hand_p0, next_id = create_hand(p0_hand_defs, PlayerId(0))
    hand_p1, next_id = create_hand(p1_hand_defs, PlayerId(1))

    player_0 = PlayerState(
        player_id=PlayerId(0),
        deck=(),
        hand=hand_p0,
        energy=Energy(p0_energy),
        max_energy=Energy(p0_energy),
    )
    player_1 = PlayerState(
        player_id=PlayerId(1),
        deck=(),
        hand=hand_p1,
        energy=Energy(p1_energy),
        max_energy=Energy(p1_energy),
    )

    return GameState(
        turn=TurnNumber(turn),
        phase=GamePhase.PLANNING,
        players=(player_0, player_1),
        locations=create_initial_locations(),
        result=GameResult.IN_PROGRESS,
        next_instance_id=next_id,
    )

File: web/src/engine/controller.ts
/**
 * Game Controller - orchestrates game phases
 * 
 * This module provides two resolution modes:
 * - Legacy: resolveTurn() - sequential processing (backward compatible)
 * - Deterministic: resolveTurnDeterministic() - timeline-based pre-computation
 * 
 * For new features and Firebase/PvP, use the deterministic mode.
 */

import type {
  GameState,
  PlayerState,
  PlayerAction,
  PlayCardAction,
} from './models';
import {
  createInitialLocations,
  getPlayer,
  getLocation,
  withPlayer,
  withLocation,
  withTurn,
  withPhase,
  withResult,
  drawCard,
  drawCardWeighted,
  spendEnergy,
  removeFromHand,
  addCard,
  getCardCount,
  getTotalPower,
  clearTurnTracking,
  getBonusEnergyNextTurn,
  clearBonusEnergyNextTurn,
} from './models';
import type { GameEvent } from './events';
import type { PlayerId, LocationIndex, TurnNumber } from './types';
import { MAX_TURNS, LOCATION_CAPACITY, STARTING_HAND_SIZE, MAX_HAND_SIZE, isValidLocationIndex } from './types';
import { getDeckCardDefs, createDeck, shuffleDeckByCost, getCardDefsFromIds, getCardsByCostMap, getAllCardDefs } from './cards';
import type { CardDef } from './models';
import type { CardId } from './types';

// Timeline-based deterministic resolution imports
import { SeededRNG, generateGameSeed } from './rng';
import { generateTimeline } from './timeline/generator';
import { executeTimeline, createStepIterator } from './timeline/executor';
import type { PlayedCard, ResolutionTimeline } from './timeline/types';

// =============================================================================
// Game Creation
// =============================================================================

export function createGame(): { state: GameState; events: GameEvent[] } {
  const events: GameEvent[] = [];

  // Create decks for both players (shuffled by cost - low cost cards drawn first)
  const p0Defs = shuffleDeckByCost(getDeckCardDefs('starter'));
  const p1Defs = shuffleDeckByCost(getDeckCardDefs('starter'));

  const { deck: deck0, nextId: nextId0 } = createDeck(p0Defs, 0, 0);
  const { deck: deck1, nextId: nextId1 } = createDeck(p1Defs, 1, nextId0);

  // Create player states
  let player0: PlayerState = {
    playerId: 0,
    deck: deck0,
    hand: [],
    energy: 0,
    maxEnergy: 0,
  };

  let player1: PlayerState = {
    playerId: 1,
    deck: deck1,
    hand: [],
    energy: 0,
    maxEnergy: 0,
  };

  // Draw initial hands
  for (let i = 0; i < STARTING_HAND_SIZE; i++) {
    const [newP0, card0] = drawCard(player0);
    player0 = newP0;
    if (card0) {
      events.push({ type: 'CardDrawn', playerId: 0, cardInstanceId: card0.instanceId });
    }

    const [newP1, card1] = drawCard(player1);
    player1 = newP1;
    if (card1) {
      events.push({ type: 'CardDrawn', playerId: 1, cardInstanceId: card1.instanceId });
    }
  }

  const state: GameState = {
    turn: 1 as TurnNumber,
    phase: 'PLANNING',
    players: [player0, player1],
    locations: createInitialLocations(),
    result: 'IN_PROGRESS',
    nextInstanceId: nextId1,
    cardsDestroyedThisGame: [],
    cardsMovedThisGame: [],
    cardsMovedThisTurn: [],
    silencedCards: [],
    bonusEnergyNextTurn: [0, 0],
  };

  // Set energy for turn 1
  const startingEnergy = 1;
  const stateWithEnergy = withPlayer(
    withPlayer(state, 0, { ...player0, energy: startingEnergy, maxEnergy: startingEnergy }),
    1,
    { ...player1, energy: startingEnergy, maxEnergy: startingEnergy }
  );

  events.push({ type: 'GameStarted' });
  events.push({ type: 'TurnStarted', turn: 1 as TurnNumber });
  events.push({ type: 'EnergySet', playerId: 0, energy: startingEnergy });
  events.push({ type: 'EnergySet', playerId: 1, energy: startingEnergy });

  return { state: stateWithEnergy, events };
}

// =============================================================================
// Turn Management
// =============================================================================

/**
 * Count how many locations a player is currently winning.
 * A player wins a location if they have strictly more power than the opponent.
 */
function countLocationsWon(state: GameState, playerId: PlayerId): number {
  const enemyId = (1 - playerId) as PlayerId;
  let count = 0;

  for (const location of state.locations) {
    const playerPower = getTotalPower(location, playerId);
    const enemyPower = getTotalPower(location, enemyId);
    if (playerPower > enemyPower) {
      count++;
    }
  }

  return count;
}

export function startNextTurn(state: GameState): { state: GameState; events: GameEvent[] } {
  const events: GameEvent[] = [];

  const newTurn = (state.turn + 1) as TurnNumber;
  if (newTurn > MAX_TURNS) {
    // Game should be over
    return { state, events };
  }

  let newState = withTurn(state, newTurn);
  newState = withPhase(newState, 'PLANNING');
  newState = clearTurnTracking(newState);

  events.push({ type: 'TurnStarted', turn: newTurn });

  // Set energy and draw cards
  for (const playerId of [0, 1] as PlayerId[]) {
    let player = getPlayer(newState, playerId);

    // Base energy = turn number
    const baseEnergy = newTurn;

    // Bonus energy from locations: +1 for each location currently won
    // Use the state from END of previous turn (before this turn's updates)
    const locationsWon = countLocationsWon(state, playerId);
    const locationBonus = locationsWon;

    // Bonus energy from card effects (e.g., Iris)
    const cardEffectBonus = getBonusEnergyNextTurn(state, playerId);

    const totalBonus = locationBonus + cardEffectBonus;
    const totalEnergy = baseEnergy + totalBonus;

    player = { ...player, energy: totalEnergy, maxEnergy: totalEnergy };
    newState = withPlayer(newState, playerId, player);
    events.push({ type: 'EnergySet', playerId, energy: baseEnergy });

    // Emit bonus energy event if player gets bonus from locations
    if (locationBonus > 0) {
      events.push({
        type: 'BonusEnergy',
        playerId,
        bonus: locationBonus,
        locationsWon,
        newTotal: totalEnergy,
      });
    }

    // Emit bonus energy event if player gets bonus from card effects
    if (cardEffectBonus > 0) {
      events.push({
        type: 'BonusEnergy',
        playerId,
        bonus: cardEffectBonus,
        locationsWon: 0, // From card effect, not locations
        newTotal: totalEnergy,
      });
    }

    // Draw cards to fill hand to 4 cards (weighted towards higher cost cards on later turns)
    const TARGET_HAND_SIZE = 4;
    while (player.hand.length < TARGET_HAND_SIZE && player.hand.length < MAX_HAND_SIZE) {
      const [drawnPlayer, card] = drawCardWeighted(player, newTurn);
      if (card) {
        player = drawnPlayer;
        newState = withPlayer(newState, playerId, player);
        events.push({ type: 'CardDrawn', playerId, cardInstanceId: card.instanceId });
      } else {
        // No more cards in deck
        break;
      }
    }
  }

  // Clear bonus energy from card effects (it's been consumed)
  newState = clearBonusEnergyNextTurn(newState);

  return { state: newState, events };
}

// =============================================================================
// Action Validation
// =============================================================================

export interface ValidationResult {
  valid: boolean;
  reason: string;
}

export function validateAction(state: GameState, action: PlayerAction): ValidationResult {
  if (action.type === 'Pass') {
    return { valid: true, reason: '' };
  }

  const { playerId, cardInstanceId, location } = action;

  // Check location is valid
  if (!isValidLocationIndex(location)) {
    return { valid: false, reason: `Invalid location: ${location}` };
  }

  const player = getPlayer(state, playerId);
  const loc = getLocation(state, location);

  // Check card is in hand
  const card = player.hand.find(c => c.instanceId === cardInstanceId);
  if (!card) {
    return { valid: false, reason: 'Card not in hand' };
  }

  // Check energy
  if (card.cardDef.cost > player.energy) {
    return { valid: false, reason: `Not enough energy: need ${card.cardDef.cost}, have ${player.energy}` };
  }

  // Check location capacity
  if (getCardCount(loc, playerId) >= LOCATION_CAPACITY) {
    return { valid: false, reason: `Location ${location} is at capacity` };
  }

  return { valid: true, reason: '' };
}

export function getLegalActions(state: GameState, playerId: PlayerId): PlayerAction[] {
  const actions: PlayerAction[] = [{ type: 'Pass', playerId }];

  const player = getPlayer(state, playerId);
  for (const card of player.hand) {
    for (const locIdx of [0, 1, 2] as LocationIndex[]) {
      const action: PlayCardAction = {
        type: 'PlayCard',
        playerId,
        cardInstanceId: card.instanceId,
        location: locIdx,
      };
      if (validateAction(state, action).valid) {
        actions.push(action);
      }
    }
  }

  return actions;
}

// =============================================================================
// NOTE: Legacy resolveTurn() and its helper functions have been removed.
// All turn resolution now uses resolveTurnDeterministic() which provides
// deterministic behavior through the timeline-based ability system.
// See web/src/engine/timeline/ and web/src/engine/ability/ for the new system.
// =============================================================================

// =============================================================================
// Win Condition
// =============================================================================

export function computeWinner(state: GameState): {
  result: import('./types').GameResult;
  locationWinners: (PlayerId | null)[];
  locationPowers: [number, number][];
  totalPower: [number, number];
} {
  const locationWinners: (PlayerId | null)[] = [];
  const locationPowers: [number, number][] = [];
  let p0Wins = 0;
  let p1Wins = 0;
  let totalP0 = 0;
  let totalP1 = 0;

  for (const location of state.locations) {
    const p0Power = getTotalPower(location, 0);
    const p1Power = getTotalPower(location, 1);
    locationPowers.push([p0Power, p1Power]);
    totalP0 += p0Power;
    totalP1 += p1Power;

    if (p0Power > p1Power) {
      locationWinners.push(0);
      p0Wins++;
    } else if (p1Power > p0Power) {
      locationWinners.push(1);
      p1Wins++;
    } else {
      locationWinners.push(null);
    }
  }

  let result: import('./types').GameResult;
  if (p0Wins >= 2) {
    result = 'PLAYER_0_WINS';
  } else if (p1Wins >= 2) {
    result = 'PLAYER_1_WINS';
  } else if (totalP0 > totalP1) {
    result = 'PLAYER_0_WINS';
  } else if (totalP1 > totalP0) {
    result = 'PLAYER_1_WINS';
  } else {
    result = 'DRAW';
  }

  return { result, locationWinners, locationPowers, totalPower: [totalP0, totalP1] };
}

// =============================================================================
// Deterministic Turn Resolution (Timeline-based)
// =============================================================================

/**
 * Result of deterministic turn resolution.
 */
export interface DeterministicResolutionResult {
  /** Final game state after resolution */
  readonly state: GameState;
  /** All events emitted during resolution */
  readonly events: GameEvent[];
  /** The pre-computed resolution timeline */
  readonly timeline: ResolutionTimeline;
  /** Whether resolution completed successfully */
  readonly success: boolean;
  /** Error message if resolution failed */
  readonly error?: string;
}

/**
 * Resolve a turn using the deterministic timeline system.
 * 
 * This is the new resolution mode that pre-computes all events before
 * executing them. Use this for:
 * - Firebase/PvP games (identical results across clients)
 * - Replays (same inputs = same outputs)
 * - Testing (deterministic assertions)
 * 
 * Phase A (COMMIT):
 *   - Cards are placed face-down at locations
 *   - No abilities trigger
 * 
 * Phase B (RESOLUTION):
 *   - Timeline is generated (pre-computed)
 *   - Timeline is executed step-by-step
 *   - Events are emitted for UI animation
 * 
 * Phase C (STABILIZATION):
 *   - Ongoing abilities are recalculated
 *   - Temporary effects expire
 *   - Location control is recalculated
 * 
 * @param state - Current game state
 * @param action0 - Player 0's action
 * @param action1 - Player 1's action
 * @param rng - Seeded RNG (optional - will be created from game state if not provided)
 */
export function resolveTurnDeterministic(
  state: GameState,
  action0: PlayerAction,
  action1: PlayerAction,
  rng?: SeededRNG
): DeterministicResolutionResult {
  const events: GameEvent[] = [];
  
  // Create RNG if not provided
  const effectiveRng = rng ?? new SeededRNG(generateGameSeed('game', state.turn));
  
  // ==========================================================================
  // PHASE A: COMMIT
  // Place cards face-down at locations. No abilities trigger.
  // ==========================================================================
  
  let newState = withPhase(state, 'RESOLUTION');
  const playedCards: PlayedCard[] = [];
  let playOrder = 0;
  
  // Process both actions - place cards face-down
  for (const action of [action0, action1]) {
    const validation = validateAction(newState, action);
    if (!validation.valid) {
      events.push({ type: 'ActionInvalid', playerId: action.playerId, reason: validation.reason });
      continue;
    }
    
    if (action.type === 'Pass') {
      events.push({ type: 'PlayerPassed', playerId: action.playerId });
      continue;
    }
    
    // PlayCard action
    const player = getPlayer(newState, action.playerId);
    const [playerWithoutCard, card] = removeFromHand(player, action.cardInstanceId);
    
    if (!card) continue;
    
    // Spend energy
    const playerSpent = spendEnergy(playerWithoutCard, card.cardDef.cost);
    newState = withPlayer(newState, action.playerId, playerSpent);
    events.push({
      type: 'EnergySpent',
      playerId: action.playerId,
      amount: card.cardDef.cost,
      remaining: playerSpent.energy,
    });
    
    // Place card at location (unrevealed)
    const location = getLocation(newState, action.location);
    const newLocation = addCard(location, card, action.playerId);
    newState = withLocation(newState, action.location, newLocation);
    
    events.push({
      type: 'CardPlayed',
      playerId: action.playerId,
      cardInstanceId: card.instanceId,
      location: action.location,
    });
    
    // Track played card for timeline generation
    playedCards.push({
      instanceId: card.instanceId,
      playerId: action.playerId,
      location: action.location,
      playOrder: playOrder++,
    });
  }
  
  // ==========================================================================
  // PHASE B: RESOLUTION
  // Generate timeline (pre-computed) and execute it.
  // ==========================================================================
  
  // Emit resolution started event
  events.push({
    type: 'ResolutionStarted',
    turn: state.turn as TurnNumber,
    totalSteps: 0, // Will be updated after generation
  } as GameEvent);
  
  // Generate the resolution timeline
  const timelineResult = generateTimeline(newState, playedCards, effectiveRng);
  
  // Update resolution started event with actual step count
  const resolutionStartedIdx = events.findIndex(e => e.type === 'ResolutionStarted');
  if (resolutionStartedIdx >= 0) {
    events[resolutionStartedIdx] = {
      type: 'ResolutionStarted',
      turn: state.turn as TurnNumber,
      totalSteps: timelineResult.timeline.length,
    } as GameEvent;
  }
  
  // Execute the timeline
  const executionResult = executeTimeline(newState, timelineResult.timeline, effectiveRng);
  newState = executionResult.state;
  events.push(...executionResult.events);
  
  // Emit resolution ended event
  events.push({
    type: 'ResolutionEnded',
    turn: state.turn as TurnNumber,
  } as GameEvent);
  
  // ==========================================================================
  // PHASE C: STABILIZATION
  // Check for game end and transition to next phase.
  // ==========================================================================
  
  // Check for game end
  if (newState.turn >= MAX_TURNS) {
    const { result, locationWinners, locationPowers, totalPower } = computeWinner(newState);
    newState = withResult(newState, result);
    newState = withPhase(newState, 'GAME_OVER');
    events.push({
      type: 'GameEnded',
      result,
      locationWinners,
      locationPowers,
      totalPower,
    });
  } else {
    newState = withPhase(newState, 'TURN_END');
    events.push({ type: 'TurnEnded', turn: newState.turn });
  }
  
  return {
    state: newState,
    events,
    timeline: timelineResult.timeline,
    success: executionResult.success,
    error: executionResult.error,
  };
}

// =============================================================================
// NPC Deck Generation
// =============================================================================

/**
 * Number of card swaps based on difficulty/unlock position.
 * Lower positions = easier = fewer swaps (more similar to player's deck)
 */
function getNpcSwapCount(unlockPosition: number): number {
  if (unlockPosition < 5) return 2;      // Easy: 2 random swaps
  if (unlockPosition < 10) return 4;     // Medium-easy: 4 swaps
  if (unlockPosition < 15) return 6;     // Medium: 6 swaps
  if (unlockPosition < 25) return 8;     // Medium-hard: 8 swaps
  return 10;                              // Hard: 10 swaps (more variety)
}

/**
 * Generate NPC deck based on player's deck with random same-cost card swaps.
 * 
 * This creates a fair matchup where:
 * - NPC has the same deck size as the player
 * - NPC has similar card power distribution (same costs)
 * - NPC has some variety through random swaps
 * - Swap count increases with player progression (harder = more variety)
 * 
 * @param playerDeckIds - The player's current deck card IDs
 * @param swapCount - Number of cards to randomly replace
 * @param rng - Seeded RNG for deterministic swaps
 * @returns Array of CardDefs for the NPC's deck
 */
function generateNpcDeck(
  playerDeckIds: CardId[],
  swapCount: number,
  rng: SeededRNG
): CardDef[] {
  // Get player's deck definitions
  const playerDeck = getCardDefsFromIds(playerDeckIds);
  if (playerDeck.length === 0) {
    // Fallback to starter deck if player deck is empty
    return getDeckCardDefs('starter');
  }

  // Get all cards grouped by cost for finding replacements
  const cardsByCost = getCardsByCostMap();
  const playerCardIds = new Set(playerDeckIds);
  
  // Start with a copy of the player's deck
  const npcDeck = [...playerDeck];
  
  // Track which indices we've already swapped to avoid double-swapping
  const swappedIndices = new Set<number>();
  
  // Perform swaps
  let swapsPerformed = 0;
  let attempts = 0;
  const maxAttempts = swapCount * 3; // Prevent infinite loops
  
  while (swapsPerformed < swapCount && attempts < maxAttempts) {
    attempts++;
    
    // Pick a random card from the deck to swap
    const indexToSwap = Math.floor(rng.next() * npcDeck.length);
    
    // Skip if already swapped this index
    if (swappedIndices.has(indexToSwap)) continue;
    
    const cardToReplace = npcDeck[indexToSwap]!;
    const cost = cardToReplace.cost;
    
    // Find all cards with the same cost that aren't in the current NPC deck
    const currentNpcCardIds = new Set(npcDeck.map(c => c.id));
    const sameCostCards = cardsByCost.get(cost) ?? [];
    const availableReplacements = sameCostCards.filter(
      card => !currentNpcCardIds.has(card.id)
    );
    
    if (availableReplacements.length === 0) continue;
    
    // Pick a random replacement
    const replacementIndex = Math.floor(rng.next() * availableReplacements.length);
    const replacement = availableReplacements[replacementIndex]!;
    
    // Perform the swap
    npcDeck[indexToSwap] = replacement;
    swappedIndices.add(indexToSwap);
    swapsPerformed++;
  }
  
  return npcDeck;
}

/**
 * Create a game with a specific seed for deterministic behavior.
 * 
 * @param seed - Random seed for deterministic behavior
 * @param playerDeckIds - Optional array of card IDs for player 0's deck (uses starter deck if not provided)
 * @param unlockPosition - Player's unlock position for difficulty scaling (defaults to 0)
 */
export function createGameWithSeed(
  seed: number,
  playerDeckIds?: CardId[],
  unlockPosition: number = 0
): { state: GameState; events: GameEvent[]; rng: SeededRNG } {
  const rng = new SeededRNG(seed);
  const events: GameEvent[] = [];
  
  // Create decks for both players (shuffled by cost - low cost cards drawn first)
  // Player 0 uses the provided deck IDs or falls back to starter deck
  const effectivePlayerDeckIds = playerDeckIds && playerDeckIds.length > 0
    ? playerDeckIds
    : getDeckCardDefs('starter').map(c => c.id);
  
  const p0Defs = rng.shuffle(getCardDefsFromIds(effectivePlayerDeckIds));
  
  // NPC deck: based on player's deck with random same-cost swaps
  // Swap count increases with player progression for more variety/challenge
  const swapCount = getNpcSwapCount(unlockPosition);
  const p1Defs = rng.shuffle(generateNpcDeck(effectivePlayerDeckIds, swapCount, rng));
  
  // Sort by cost for early-game playability
  p0Defs.sort((a, b) => a.cost - b.cost);
  p1Defs.sort((a, b) => a.cost - b.cost);
  
  const { deck: deck0, nextId: nextId0 } = createDeck(p0Defs, 0, 0);
  const { deck: deck1, nextId: nextId1 } = createDeck(p1Defs, 1, nextId0);
  
  // Create player states
  let player0: PlayerState = {
    playerId: 0,
    deck: deck0,
    hand: [],
    energy: 0,
    maxEnergy: 0,
  };
  
  let player1: PlayerState = {
    playerId: 1,
    deck: deck1,
    hand: [],
    energy: 0,
    maxEnergy: 0,
  };
  
  // Draw initial hands
  for (let i = 0; i < STARTING_HAND_SIZE; i++) {
    const [newP0, card0] = drawCard(player0);
    player0 = newP0;
    if (card0) {
      events.push({ type: 'CardDrawn', playerId: 0, cardInstanceId: card0.instanceId });
    }
    
    const [newP1, card1] = drawCard(player1);
    player1 = newP1;
    if (card1) {
      events.push({ type: 'CardDrawn', playerId: 1, cardInstanceId: card1.instanceId });
    }
  }
  
  const state: GameState = {
    turn: 1 as TurnNumber,
    phase: 'PLANNING',
    players: [player0, player1],
    locations: createInitialLocations(),
    result: 'IN_PROGRESS',
    nextInstanceId: nextId1,
    cardsDestroyedThisGame: [],
    cardsMovedThisGame: [],
    cardsMovedThisTurn: [],
    silencedCards: [],
    bonusEnergyNextTurn: [0, 0],
  };
  
  // Set energy for turn 1
  const startingEnergy = 1;
  const stateWithEnergy = withPlayer(
    withPlayer(state, 0, { ...player0, energy: startingEnergy, maxEnergy: startingEnergy }),
    1,
    { ...player1, energy: startingEnergy, maxEnergy: startingEnergy }
  );
  
  events.push({ type: 'GameStarted' });
  events.push({ type: 'TurnStarted', turn: 1 as TurnNumber });
  events.push({ type: 'EnergySet', playerId: 0, energy: startingEnergy });
  events.push({ type: 'EnergySet', playerId: 1, energy: startingEnergy });
  
  return { state: stateWithEnergy, events, rng };
}

/**
 * Get a step-by-step iterator for animating resolution.
 * 
 * This allows the UI to execute one step at a time with animation pauses.
 * 
 * @param state - Game state after Phase A (cards placed)
 * @param timeline - Pre-computed resolution timeline
 * @param rng - Seeded RNG
 */
export function createResolutionIterator(
  state: GameState,
  timeline: ResolutionTimeline,
  rng: SeededRNG
) {
  return createStepIterator(state, timeline, rng);
}
